source/common/upstream/cluster_manager_impl.cc

```diff
@@ -739,11 +763,29 @@ void ClusterManagerImpl::ThreadLocalClusterManagerImpl::onHostHealthFailure(
   // more granular host set changes, we should be able to capture single host changes and make them
   // more targeted.
   ThreadLocalClusterManagerImpl& config = tls.getTyped<ThreadLocalClusterManagerImpl>();
-  const auto& container = config.host_http_conn_pool_map_.find(host);
-  if (container != config.host_http_conn_pool_map_.end()) {
-    for (const auto& pair : container->second.pools_) {
-      const Http::ConnectionPool::InstancePtr& pool = pair.second;
-      pool->drainConnections();
+  {
+    const auto& container = config.host_http_conn_pool_map_.find(host);
+    if (container != config.host_http_conn_pool_map_.end()) {
+      for (const auto& pair : container->second.pools_) {
+        const Http::ConnectionPool::InstancePtr& pool = pair.second;
+        pool->drainConnections();
+      }
+    }
+  }
+
+  if (host->cluster().features() &
+      ClusterInfo::Features::CLOSE_CONNECTIONS_ON_HOST_HEALTH_FAILURE) {
+
+    // Each connection will remove itself from the TcpConnectionsMap when it closes, via it's
+    // Network::ConnectionCallbacks. The last removed tcp conn will remove the TcpConnectionsMap
+    // from host_tcp_conn_map_, so do not cache it between iterations.
+    while (true) {
+      const auto& it = config.host_tcp_conn_map_.find(host);
+      if (it == config.host_tcp_conn_map_.end()) {
+        break;
+      }
+      TcpConnectionsMap& container = it->second;
+      container.begin()->first->close(Network::ConnectionCloseType::NoFlush);
```

Comment:
E_INSUFFICIENT_KNOWLEDGE question - does this cause synchronous teardown of all the associated state of all the associated TCP connections?  If so I have concerns about doing it one fell swoop rather than either 1) making it async or 2) batching and having alarms.

I'm used to servers with hundreds of thousands of connections (sometimes though admittedly less often to a limited number of backends) so loops like this tend to scare me.  If we think implementation is overkill maybe a warning in the config field that scaling this is T.B.D (or is that overkill too?)

Reply:
Yeah, this is a scalability problem if there are large numbers of connections to each upstream instance. It's worth thinking about optimizing this in the future. The same issue exists for http connections right now; we don't close them, but we do iterate through all of them.

I'll add a note to the docs.